File: graph_algorithms/week5_spanning_trees/2_clustering/clustering.cpp
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <tuple>
#include <vector>

using std::pair;
using std::tuple;
using std::vector;

int find(vector<int>& parent, int i) {
  if (i != parent[i]) {
    parent[i] = find(parent, parent[i]);
  }
  return parent[i];
}

void unionSets(vector<int>& parent, vector<int>& rank, int i, int j) {
  int iParent = find(parent, i);
  int jParent = find(parent, j);
  if (iParent == jParent)
    return;

  if (rank[iParent] > rank[jParent])
    parent[jParent] = iParent;
  else {
    parent[iParent] = jParent;
    if (rank[iParent] == rank[jParent])
      rank[jParent]++;
  }
}

float min_distance_clustering(vector<tuple<int, int, float>>& edges, int n, int k) {
  vector<int> parent(n);
  vector<int> rank(n, 0);
  int nEdge = 0;

  for (int i = 0; i < n; i++)
    parent[i] = i;

  sort(edges.begin(), edges.end(), [](const tuple<int, int, float>& a, const tuple<int, int, float>& b) {
    return std::get<2>(a) < std::get<2>(b);
  });

  for (const auto& edge : edges) {
    int u, v;
    float w;
    std::tie(u, v, w) = edge;
    if (find(parent, u) != find(parent, v)) {
      if (nEdge == n - k)
        return w;

      nEdge++;
      unionSets(parent, rank, u, v);
    }
  }

  return 0.0;
}

int main() {
  int nVertices;
  std::cin >> nVertices;
  vector<pair<int, int>> points(nVertices);

  for (int i = 0; i < nVertices; i++) {
    int a, b;
    std::cin >> a >> b;
    points[i] = std::make_pair(a, b);
  }

  int nSubsets;
  std::cin >> nSubsets;

  vector<tuple<int, int, float>> edges;

  for (int i = 0; i < nVertices; i++) {
    int x0, y0;
    std::tie(x0, y0) = points[i];
    for (int j = i + 1; j < nVertices; j++) {
      int x, y;
      std::tie(x, y) = points[j];
      float distance = sqrt(pow(x - x0, 2) + pow(y - y0, 2));
      edges.push_back(std::make_tuple(i, j, distance));
    }
  }

  std::cout << std::fixed << std::setprecision(9) << min_distance_clustering(edges, nVertices, nSubsets) << std::endl;

  return 0;
}

File: graph_algorithms/week5_spanning_trees/2_clustering/clustering.py
import math
from typing import List, Optional, Tuple


def find(parent: List[int], i: int) -> int:
    if i != parent[i]:
        parent[i] = find(parent, parent[i])
    return parent[i]


def union(parent: List[int], rank: List[int], i: int, j: int) -> None:  # noqa: C901
    i_parent = find(parent, i)
    j_parent = find(parent, j)
    if i_parent == j_parent:
        return

    if rank[i_parent] > rank[j_parent]:
        parent[j_parent] = i_parent
    else:
        parent[i_parent] = j_parent
        if rank[i_parent] == rank[j_parent]:
            rank[j_parent] += 1


def min_distance_clustering(edges: List[Tuple[int, int, float]], n: int, k: int) -> float:  # noqa: C901
    parent = list(range(n))
    rank = [0] * n
    n_edge = 0
    edges.sort(key=lambda i: i[2])
    for u, v, w in edges:
        if find(parent, u) != find(parent, v):
            if n_edge == n - k:
                return w

            n_edge += 1
            union(parent, rank, u, v)
    return 0


if __name__ == "__main__":  # noqa: C901
    n_vertices = int(input())
    points: List[Optional[Tuple[int, int]]] = [None] * n_vertices  # noqa: TAE002
    for i in range(n_vertices):
        a, b = map(int, input().split())
        points[i] = (a, b)

    edges = []  # (start, end, weight)
    n_subsets = int(input())
    for i in range(n_vertices):
        (x0, y0) = points[i]  # type: ignore
        for j in range(i + 1, n_vertices):
            (x, y) = points[j]  # type: ignore
            distance = math.sqrt((x - x0) ** 2 + (y - y0) ** 2)
            edges.append((i, j, distance))

    min_d = min_distance_clustering(edges=edges, n=n_vertices, k=n_subsets)
    print("{0:.9f}".format(min_d))
